### PR TITLE
Fix numerical instability in higher-order gradients of Hafnian (#133)

### LINKED ISSUE #133: `hafnian` fails to backward if construction of derivative graph is required
- If  `torch.autograd.grad` parameter `create_graph` is set as `True`, backward of loss function(output of GBS) fails. This results in failure of derivative graph based features like `hessian`.

- Through examination, this issue relates to the `torch.linalg.eigvals` of hafnian calculation. However, `torch.linalg.eigvals` only will not cause this error, the root cause requires further investigation.

- Minimal Reproduction

1. hessian
```
from torch.autograd.functional import hessian
def f(params):
    cir = dq.QumodeCircuit(nmode=2, cutoff=3, init_state='vac', backend='gaussian')
    cir.s(wires=[0], r=0.8, theta=0)
    cir.s(wires=[1], r=0.8, theta=0)

    cir.bs([0,1])
    cir.ps([0], encode=True)
    cir.bs([0,1])

    d = cir(params, is_prob=True)
    sorted_values = [v for k, v in d.items()][:2]
    probs = torch.cat(sorted_values)
    return probs.sum()

x = torch.tensor([1.0], requires_grad=True)
hessian(f, x)
```

Image

Random matrix with `torch.linalg.eigvals` only will not cause this error:
Image

2. backward fails when `create_graph=True`
```
cir = dq.QumodeCircuit(nmode=2, cutoff=3, init_state='vac', backend='gaussian')
cir.s(wires=[0], r=0.8, theta=0)
cir.s(wires=[1], r=0.8, theta=0)

cir.bs([0,1])
cir.ps([0], encode=True)
cir.bs([0,1])

optimizer = optim.Adam(cir.parameters(), lr=0.01)
epochs = 3

for epoch in range(epochs):
    optimizer.zero_grad()

    t = torch.tensor([1.0], requires_grad=True)

    d = cir(t, is_prob = True)
    sorted_values = [v for k, v in d.items()][:2]
    u = torch.cat(sorted_values).sum()

    du_dt = torch.autograd.grad(
        outputs=u, inputs=t,
        create_graph=True, retain_graph=True
    )[0]

    loss = du_dt + u
    # torch.autograd.set_detect_anomaly(True)
    # with torch.autograd.detect_anomaly():
    #     loss.backward()
    loss.backward()
    optimizer.step()

    if epoch % 1 == 0:
        print(f"Epoch {epoch}: Loss {loss.item():.6f}")
```

Image

 In such cases, `create_graph` can only be set as `True`, otherwise the backward of loss will neglect the derivative part. (verified numerically✔)

### PR DESCRIPTION
**Pull request by Hugh-888**
- Root cause
The numerical instability comes from the trace power calculation in function ``poly_lambda``, where ``torch.eigen`` is used.
When the matrix has degenerate or nearly degenerate eigenvalues, the backward pass involves divisions by small eigenvalue gaps, which leads to ``NaN`` results in the Hessian computation.
- Solution
 We replace the eigenvalue-based trace power computation with an iterative approach. A new function ``trace_powers`` is introduced. A simple case is as follows,
image

- Simple benchmark
The benchmark results (hafnian) are shown as follows,
image

-- blue curve: the new iterative implementation ``trace_powers``
-- orange curve: Pytorch built-in ``torch.matrix_power``
-- green curve: the original eigenvalue decomposition–based approach

The benchmark results  for the photon loss case with displacement (loop hafnian),
image

The benchmark results  for hafnian calculation for matrix sizes 4, 12, and 20

image

### CODE REVIEW

**Review by sansiro77 — CHANGES_REQUESTED** [review]
直接把迭代的算法写在`poly_lambda`里就行了，不需要单独的函数，因为loop Hafnian里也可以类似优化。
`torch.diag(a.new_ones(a.shape[0]))`可以改为`a.new_ones(a.shape[-1]).diag()`

File: src/deepquantum/photonic/hafnian_.py [hunk 1]
```diff
@@ -51,19 +51,31 @@ def get_submat_haf(a: torch.Tensor, z: torch.Tensor) -> torch.Tensor:
...
+        traces.append(torch.trace(x))
+    trace_list = torch.stack(traces)
     coeff = 0
     if loop: # loop hafnian case
         v = torch.diag(submat)
-        diag_term = torch.stack([v @ torch.linalg.matrix_power(xaz, i - 1) @ x_mat @ v / 2 for i in range(1, power+1)])
+        # matrix power calculation
+        diag_term = []
+        x = xaz.new_ones(xaz.shape[-1]).diag()
+        diag_term.append(v @ x @ x_mat @ v / 2)
```
Comment (sansiro77) [inline src/deepquantum/photonic/hafnian_.py]: `x_mat @ v / 2`可以先定义成一个新的变量来复用，比如`xv`

File: src/deepquantum/photonic/hafnian_.py [hunk 2]
```diff
@@ -51,19 +51,31 @@ def get_submat_haf(a: torch.Tensor, z: torch.Tensor) -> torch.Tensor:
...
+        for _ in range(power):
```
Comment (sansiro77) [inline src/deepquantum/photonic/hafnian_.py]: 并且和原来的for循环相比，多算了一项

File: src/deepquantum/photonic/hafnian_.py [hunk 3]
```diff
@@ -51,19 +51,31 @@ def get_submat_haf(a: torch.Tensor, z: torch.Tensor) -> torch.Tensor:
     submat = a[idx][:, idx]
     return submat
 
-
```
Comment (sansiro77) [inline src/deepquantum/photonic/hafnian_.py:54]: 多删了一行

**Review by sansiro77 — CHANGES_REQUESTED** [review]
需要再比较一下随nmode增加的效率情况，比如直接用之前benchmark的代码

**Review by sansiro77 — APPROVED** [review]
Good Job!
